File: Queue/Java/CircularQueue.java
// circular queue - array implementation
class CircularQueue{
    int size;
    int arr[];
    int front =-1,rear = -1;
    public CircularQueue(int size){
        this.size = size;
        this.arr = new int[size];
    }
    public boolean isEmpty(){
        if(front == -1 ){
            return true;
        }
        return false;
    }
    public boolean isFull(){
        if(front == rear +1 || front == 0 && rear = this.size -1){
            return true;
        }
        return false;
    }
    public int peek(){
        if(this.isEmpty()){
            System.out.println("Queue Underflow");
            System.exit(1);
        }
        return arr[front];
    }
    public void insert(int x){
        if(this.isFull()){
            System.out.println("Queue Overflow");
            return;
        }
        if(front == -1){
            front = 0;
            rear = 0;
        }
        else if(rear == this.size-1){
            rear = 0;
        }
        else{
            rear = rear+1;
        }
        arr[rear] = x;
        return;
    }
    public int delete(){
        if(this.isEmpty()){
            System.out.println("Queue Underflow");
            System.exit(1);
        }
        int x = arr[front];
        if(front == rear){
            front =-1;
            rear = -1;
        }
        else if(front == this.size-1){
            front =0;
        }
        else{
            front = front +1;
        }
        return x;
    }
    public void display(){
        
    }
}
public class Main{
    public static void main(String args[]){
        CircularQueue c = new CircularQueue(10);
        
    }
}

File: Queue/C/CircularQueue.c
// Circular Queue using array 

#include<stdio.h>
#define MAX 10

int cq[MAX];
int front =-1,rear=-1;

int isEmpty(){
    if(front == -1){
        return 1;
    }
    else{
        return 0;
    }
}

int isFull(){
    if(front == rear+1 || (front == 0 && rear ==  MAX-1)){
        return 1;
    }
    else{
        return 0;
    }
}

void display(){
    if(isEmpty()){
        printf("Queue Underflow\n");
        return;
    }
    int i = front;
    if(front < rear){
        while(i<=rear){
            printf("%d ",cq[i]);
            i += 1;
        }
    }
    else{
        while(i<=MAX-1){
            printf("%d ",cq[i]);
            i+=1;
        }
        i = 0;
        while(i<=rear){
            printf("%d ",cq[i]);
            i+=1;
        }
    }
    printf("\n");
    return;
}

void insert(int x){
    if(isFull()){
        printf("Queue Overflow\n");
        return;
    }
    if(front = -1){  // queue is empty
        front = 0;
    }
    if(rear == MAX-1){   // rear is at the last element of the array
        rear = 0;
    }
    else{
        rear = rear+1;
    }
    cq[rear] = x;
    return;
}


int delete(){
    if(isEmpty()){
        printf("Queue Underflow\n");
        return;
    }
    int x = cq[front];
    if(front == rear){ // only one element is left
        front =-1;
        rear =-1;
    }
    else if(front == MAX-1){
        front = 0;
    }
    else{
        front += 1;
    }
    return x;
}

int peek(){
    if(isEmpty()){
        printf("Queue Underflow\n");
        exit(1);
    }
    return cq[front];
}


int main(void){
    insert(5);
    insert(10);
    insert(15);
    display();
    delete();
    display();
    printf("%d\n",peek());
    insert(20);
    insert(25);
    insert(30);
    display();
    delete();
    display();
    return 0;
    
}

